src/plugins/discover/public/application/angular/discover.js

```diff
@@ -277,8 +282,20 @@ function discoverController($element, $route, $scope, $timeout, $window, Promise
   });
 
   $scope.setIndexPattern = async (id) => {
-    await replaceUrlAppState({ index: id });
-    $route.reload();
+    const nextIndexPattern = await indexPatterns.get(id);
+    if (nextIndexPattern) {
+      const nextColumns = $scope.state.columns.filter((column) =>
+        nextIndexPattern.fields.getByName(column)
+      );
```

Comment:
@timroes this would filter out all columns not available in the next index pattern selected ... but I think we should keep columns that are in neither of the both index pattern, unindexed columns?